File: src/main/java/com/yupi/airouter/service/QuotaService.java
package com.yupi.airouter.service;

/**
 * 用户配额服务
 *
 * @author <a href="https://codefather.cn">编程导航学习圈</a>
 */
public interface QuotaService {

    /**
     * 检查用户配额是否充足
     *
     * @param userId 用户ID
     * @return true-配额充足，false-配额不足
     */
    boolean checkQuota(Long userId);

    /**
     * 扣减用户Token配额
     *
     * @param userId 用户ID
     * @param tokens 消耗的Token数量
     * @return true-扣减成功，false-扣减失败
     */
    boolean deductTokens(Long userId, int tokens);

    /**
     * 获取用户剩余配额
     *
     * @param userId 用户ID
     * @return 剩余配额，-1表示无限制
     */
    long getRemainingQuota(Long userId);
}

File: python-backend/app/services/quota_service.py
"""Quota service."""

from __future__ import annotations

from sqlalchemy import select
from sqlalchemy.ext.asyncio import AsyncSession

from app.core.constants import UNLIMITED_QUOTA
from app.models.user import User


class QuotaService:
    def __init__(self, db: AsyncSession) -> None:
        self.db = db

    async def check_quota(self, user_id: int | None) -> bool:
        if user_id is None:
            return True
        user = await self.db.scalar(select(User).where(User.id == user_id, User.is_delete == 0))
        if user is None:
            return False
        if user.token_quota in (None, UNLIMITED_QUOTA):
            return True
        return (user.used_tokens or 0) < user.token_quota

    async def deduct_tokens(self, user_id: int | None, tokens: int) -> bool:
        if user_id is None or tokens <= 0:
            return True
        user = await self.db.scalar(select(User).where(User.id == user_id, User.is_delete == 0))
        if user is None:
            return False
        user.used_tokens = (user.used_tokens or 0) + tokens
        await self.db.commit()
        return True

    async def get_remaining_quota(self, user_id: int | None) -> int:
        if user_id is None:
            return UNLIMITED_QUOTA
        user = await self.db.scalar(select(User).where(User.id == user_id, User.is_delete == 0))
        if user is None:
            return 0
        if user.token_quota in (None, UNLIMITED_QUOTA):
            return UNLIMITED_QUOTA
        return max(0, user.token_quota - (user.used_tokens or 0))
